packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -405,52 +434,23 @@ export class Pipeline extends PipelineBase {
   }
 
   private ensureReplicationResourcesExistFor(action: IAction): CrossRegionInfo {
-    const pipelineStack = Stack.of(this);
+    const actionEnv = new RichAction(action).env;
 
-    let actionRegion: string | undefined;
-    let otherStack: Stack | undefined;
-
-    const actionResource = action.actionProperties.resource;
-    if (actionResource) {
-      const pipelineAndActionRegionComparison = Token.compareStrings(this.env.region, actionResource.env.region);
-      const pipelineAndActionInDifferentRegions = pipelineAndActionRegionComparison === TokenComparison.ONE_UNRESOLVED ||
-        pipelineAndActionRegionComparison === TokenComparison.DIFFERENT;
-      if (pipelineAndActionInDifferentRegions) {
-        actionRegion = actionResource.env.region;
-
-        // if the resource is from a different stack in another region but the same account,
-        // use that stack as home for the cross-region support resources
-        const actionResourceStack = Stack.of(actionResource);
-        const actionResourceAndItsStackRegionComparison = Token.compareStrings(actionResource.env.region, actionResourceStack.region);
-        const actionResourceInSameRegionAsItsStack = actionResourceAndItsStackRegionComparison === TokenComparison.SAME ||
-          actionResourceAndItsStackRegionComparison === TokenComparison.BOTH_UNRESOLVED;
-        const pipelineAndActionResourceStackAccountComparison = Token.compareStrings(this.env.account, actionResourceStack.account);
-        const pipelineAndActionResourceStackInSameAccount = pipelineAndActionResourceStackAccountComparison === TokenComparison.SAME ||
-          pipelineAndActionResourceStackAccountComparison === TokenComparison.BOTH_UNRESOLVED;
-        if (pipelineAndActionResourceStackInSameAccount && actionResourceInSameRegionAsItsStack) {
-          otherStack = actionResourceStack;
-        }
-      }
-    } else {
-      actionRegion = action.actionProperties.region;
-    }
+    // Whether the action seems to care about its account/region (or is happy
+    // to inherit from the pipeline)
+    const hasRegion = !Token.isUnresolved(actionEnv.region);
 
-    // if actionRegion is undefined,
-    // it means the action is in the same region as the pipeline -
-    // so, just return the artifactBucket
-    if (!actionRegion) {
-      return {
-        artifactBucket: this.artifactBucket,
-      };
-    }
-    // get the region the Pipeline itself is in
-    const pipelineRegion = this.requireRegion();
-    // if the action is in the same region as the pipeline, nothing to do
-    if (actionRegion === pipelineRegion) {
+    // if actionRegion is undefined or the same as the pipeline region,
+    // it means the action is in the same region as the pipeline - so, just
+    // return the artifactBucket
+    if (!hasRegion || sameEnvDimension(this.env.region, actionEnv.region)) {
```

Comment:
I still feel like this is weird. If `!hasRegion` is true, that means `Token.isUnresolved(actionRegion)` is true; so, if that's false, then there's no point in calling `sameEnvDimension()`, which explicitly checks for BOTH_UNRESOLVED, as that's simply impossible.

So, I *think* this should be changed to

```suggestion
    if (Token.isUnresolved(actionRegion) || Token.compareStrings (this.env.region, actionRegion) === TokenComparison.SAME) {
```

(It also makes the below comment:

```ts
    // (This implicitly requires the action to have a known region as well)
```

obvious too, because it's clear that `Token.isUnresolved(actionRegion)` must be false if we move past this `if`)